### ISSUE TITLE
prediction on new category should revert to grand mean

### ISSUE THREAD

**Issue by zkurtz** [microsoft/LightGBM#6908]
## Summary

Consider a very simple regression task with a single categorical (`pandas.Categorical`) predictor and a continuous output variable. Suppose, for simplicity, that there is no noise: every level of the predictor maps to a  fixed value of the target, something like

```
red -> 1
red -> 1
green -> 2
green -> 2
NULL -> 3
NULL -> 3
```

Now suppose totally overfit on this data and then ask lightgbm to predict the value for `blue`.  I'm observing that the output value in that case is "3", the same as for NULL. What makes more sense to me is if the output value would be "2", or the grand mean of the output over the training data. XGBoost adopts this latter behavior, for instance.

## Motivation

New values are different than NULL values. We learn about nulls during training. There is no a priori obvious reason that a new value should be more closely related to NULL than any other value. Therefore it would be nice at least to have the option to treat new values as uninformative, reverting to the grand mean for predictions.

## Description

I'm happy to follow up with code demoing the behavior described above if anyone cares to look more closely at a possible upgrade.

**Comment by jameslamb** [comment]
Thanks for using LightGBM and for taking the time to write this up.

> _There is no a priori obvious reason that a new value should be more closely related to NULL than any other value._

Tree-based models work by splitting the training data into subsets based on feature values.

LightGBM splits categorical features into two groupings, with rules like `feat in {red,yellow}`. Samples where the value of that feature fall in that `==` group go down one side of the split (and get the corresponding leaf value), all other samples (which includes NaN, NULL, and not-seen-during-training values) go down the other side.

This is described a bit in https://lightgbm.readthedocs.io/en/latest/Features.html#optimal-split-for-categorical-features

Given that, I think you could make a similarly-strong argument there's no "obvious" reason that a not-seen-during-training value of a categorical feature should be associated with **either** side of the split.

@guolinke gave a very good summary of this in https://github.com/microsoft/LightGBM/issues/2921 (and could correct any mistakes I've made here, if he has time).

> _What makes more sense to me is if the output value would be "2", or the grand mean of the output over the training data._

For trees with more than a 1 split, how would the behavior "revert to the grand mean for predictions" work? For example, are you saying that you'd want LightGBM to skip *all* trees in the model and instead predict a "grand mean" if *any* categorical features in the input had values that were unseen during training time?

If that is what you're looking for, I would not support adding that to LightGBM. I think that'd better be handled by preprocessing outside of LightGBM.

This is a good doc on how tree-based models work, you might find it helpful as a reference in this discussion: https://xgboost.readthedocs.io/en/stable/tutorials/model.html

> _XGBoost adopts this latter behavior_

Are you able to point to some documentation or demonstrate with a minimal, reproducible example that this is what XGBoost does?

In https://xgboost.readthedocs.io/en/stable/faq.html#how-to-deal-with-missing-values, it says "In tree algorithms, branch directions for missing values are learned during training". 

I tried tonight with `lightgbm==4.6.0` and `xgboost==3.0.1`. In both cases, for regression models, the models predicted the same output value for NaN values of a categorical feature as they did for not-seen-during training values.

<details><summary>example code (click me)</summary>

Installed like this:

```shell
conda install -c conda-forge \
  'lightgbm>=4.6.0' \
  'xgboost>=3.0'
```

Then ran the following Python code:

```python
import numpy as np
import xgboost as xgb
from sklearn.datasets import load_diabetes

# create a data set with a single feature, "bmi",
# with values case to integer
X_df, y = load_diabetes(return_X_y=True, as_frame=True, scaled=False)
X_df = X_df[["bmi"]]
X_df["bmi"] = X_df["bmi"].astype("int32")

# replace BMI=24 with NA
X_df[X_df["bmi"].isin({24,})] = np.nan

print(f"global mean of y: {np.mean(y):.6f}")
# global mean of y: 152.133484

# xgboost model (with 'bmi' treated as categorical)
xgb_model = xgb.train(
    params={
        "max_depth": 1,
        "min_child_weight": 1e-6,
        "nthread": 1,
        "objective": "reg:squarederror",
        "seed": 708,
    },
    dtrain=xgb.DMatrix(
        data=X_df,
        label=y,
        feature_types=["c"]
    ),
    num_boost_round=1,
)

# predict on the value '24' (not seen during training)
xgb_model.predict(
    data=xgb.DMatrix(
        data=np.array([24], dtype=np.int32).reshape(1, 1),
        feature_names=["bmi"],
        feature_types=["c"],
    ),
)
# array([142.51573], dtype=float32

# predict on NaN
xgb_model.predict(
    data=xgb.DMatrix(
        data=np.array([np.nan]).reshape(1, 1),
        feature_names=["bmi"],
        feature_types=["c"],
    ),
)
# array([142.51573], dtype=float32)

import lightgbm as lgb

# lightgbm model (with 'bmi' treated as categorical)
lgb_model = lgb.train(
    params={
        "min_data_in_bin": 1,
        "min_data_in_leaf": 1,
        "max_depth": 1,
        "num_threads": 1,
        "objective": "regression",
        "seed": 708,
    },
    train_set=lgb.Dataset(
        data=X_df,
        label=y,
        categorical_feature=["bmi"]
    ),
    num_boost_round=1,
)

# predict on the value '24' (not seen during training)
lgb_model.predict(np.array([24], dtype=np.int32).reshape(1, 1))
# array([155.03504499])

# predict on NaN
lgb_model.predict(np.array([np.nan]).reshape(1, 1))
# array([155.03504499])
```

</details>

**Comment by zkurtz** [comment]
Thanks for looking at this. I'll break this down into three parts:
- clarify what I'm looking for in terms of tree behavior
- explain where your example differs from the premise of my feature request
- provide a code example where xgboost behavior is consistent with my request

## Desired behavior

> For trees with more than a 1 split, how would the behavior "revert to the grand mean for predictions" work? For example, are you saying that you'd want LightGBM to skip all trees in the model and instead predict a "grand mean" if any categorical features in the input had values that were unseen during training time?

Saying "grand mean" was probably misleading on my part. Suppose we're passing a data point (row, sample, etc) through a tree and we come to a node that defines a split on our categorical feature. My suggestion is that if the value of that feature was not observed during training, tree traversal would end there, and the returned prediction would be the mean of all training observations that passed through that node. In other words, we'd treat that node as if it were a leaf rather than going all the way out to an actual leaf.

## Key premise: distinct encoding
In your example, the categorical encoding of the categorical feature is evidently happening under the hood, prompted by your use of arguments like `feature_types`. I'm willing to bet then that when the predict method sees a new category, it's literally assigning that to the same category code (`-1`) as it does for or a null value. In that sense it's completely expected that you get the same prediction for `np.nan` as you do for a not-seen value. We can make this concrete by looking at the category codes for couple of pandas categorical series:
```
pd.Series(pd.Categorical(["a", None], categories=["a"], ordered=False)).cat.codes.to_list()
# [0, -1]
pd.Series(pd.Categorical(["a", "b"], categories=["a"], ordered=False)).cat.codes.to_list()
# [0, -1]
```
In the above example, "a" is the only "known" non-null category, and the interesting bit is that both `None` and `"b"` get mapped to `-1`.

Getting to the premise of my question starts with avoiding precisely this lumping of "new" and "null" together in the same category. To do that, we need to clarify the encoding prior to passing data into lightgbm or xgboost to ensure that `None` and `"b"` each get their own category code. I do that in a careful tested way in [here](https://github.com/zkurtz/pandahandler/blob/c1b5045713ccddbb5f450b86dcca2829ede30728/pandahandler/sklearn/transforms/categoricals.py#L35-L41), but for illustration it's enough to get the idea with:
```
pd.Series(pd.Categorical(["a", "b", None], categories=["a", "b"], ordered=False)).cat.codes.to_list()
# [0, 1, -1]
```
Now `"b" -> 1`, distinct from `None -> -1`.

## Example code for xgboost

Below is an example showing xgboost delivering the behavior that I'm requesting. The mean of the labels `[1, 2, 3]` is `2`, and we have `predict(None) = 3` while `predict(new categor) = 2`:

<details>
<summary>Click to expand code</summary>
```
import pandas as pd
import xgboost as xgb

train_df = pd.DataFrame({"categorical_feature": ["category0", "category1", None]})
train_df["categorical_feature"] = train_df["categorical_feature"].astype("category")

dtrain = xgb.DMatrix(
    train_df[["categorical_feature"]],
    label=pd.Series([1, 2, 3]),
    enable_categorical=True,
)
params = {
    "objective": "reg:squarederror",
    "eta": 1.0,
    "lambda": 0,
    "alpha": 0,
    "min_child_weight": 0,
    "verbosity": 0,
}
xgb_model = xgb.train(
    params=params,
    dtrain=dtrain,
    num_boost_round=1,
)

# Now predict on data that includes a new input category:
pred_df = pd.DataFrame({"categorical_feature": ["category0", "category1", None, "category-new"]})
pred_df["categorical_feature"] = pd.Categorical(
    pred_df["categorical_feature"],
    categories=pred_df["categorical_feature"].dropna().to_list(),
    ordered=False,
)
dpred = xgb.DMatrix(pred_df, enable_categorical=True)
xgb_model.predict(dpred)
```
